### PR TITLE
Create sklearn.py

### LINKED ISSUE #307: add a scikit-learn compatible SearchCV class
Hi,

First of all, thanks for developing a nice package!

I am a heavy user of scikit-learn and often use `RandomizedSearchCV`.
So I have implemented a similar SearchCV class with Optuna.
https://github.com/Y-oHr-N/mllib/blob/master/mllib/model_selection.py

If you agree, I would like to contribute to Optuna by merging this code into your package.

Thanks.

### PR DESCRIPTION
**Pull request by Y-oHr-N**
#### Reference Issues/PRs

This PR closes #307.

#### What does this implement/fix? Explain your changes.

``TPESearchCV`` compatible with scikit-learn is added.

#### Any other comments?

TODO:

- [x] fix flake8 errors
- [x] fix mypy errors
- [x] resolve conflicts
- [x] add enable_pruning as an argument
- [x] add sampler as an argument
- [x] add tests
- [ ] add an example
- [ ] show benchmark results

### CODE REVIEW

**Comment by g-votte** [comment]
@Y-oHr-N 
Thanks for your PR! Please let us know if you have any question or concern while writing tests.

FYI: You'd need to modify several lines in `sklearn.py` and `integration/__init__.py` to pass `autopep8` check.
https://circleci.com/gh/pfnet/optuna/11067?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by Y-oHr-N** [comment]
The slight differences with ``RandomizedSearchCV`` in scikit-learn are:

- does not have ``iid``,
- does not have ``pre_dispatch``,
- does not have ``cv_results_``,
- has ``score_samples``,
- does not allow multiple metrics to be given to ``scoring``,
- comuputes CV score using ``nanmean`` instead of ``average``.

**Comment by Y-oHr-N** [comment]
@g-votte
Currently, pruning is performed whenever an estimator with ``partial_fit`` is passed. Should I modify the code to allow users to control pruning?

**Comment by g-votte** [comment]
@Y-oHr-N 

We prefer that pruning is a non-default behaviour, and activated only when the user explicitly pass some argument.
How about adding `pruning` flag as a `TPESearchCV` argument?

**Comment by Y-oHr-N** [comment]
OK.

However, ``TPESearchCV`` has ``pruner`` as a parameter. Users may be confused since they are similar.
Also, I think that the flag should be a verb (or a verb with an object) like ``load_if_exists``, ``return_train_score`` and ``refit``.

**Comment by Y-oHr-N** [comment]
According to #346, you are planning to make more sampling algorithms.
Should I modify the code so that sampler could be passed as an argument?
In that case, we need to change the class name and consider the treatment of the seed.

**Comment by sile** [comment]
@Y-oHr-N 

> However, `TPESearchCV` has `pruner` as a parameter. Users may be confused since they are similar.
Also, I think that the flag should be a verb (or a verb with an object) like `load_if_exists`, `return_train_score` and `refit`.

You're right. We think `enable_pruning` is more suitable as the name of that option. What do you think?

> Should I modify the code so that sampler could be passed as an argument?
> In that case, we need to change the class name and consider the treatment of the seed.

Nice suggestion, thanks!
It feels better to rename `TPESearchCV` to `OptunaSearchCV` and add a new `sampler` argument to the constructor.
We think that `random_state` argument can be removed if those changes are applied.

**Comment by Y-oHr-N** [comment]
> You're right. We think ``enable_pruning`` is more suitable as the name of that option. What do you think?

Sounds good. I'll add ``enable_pruning`` as an argument.

> It feels better to rename ``TPESearchCV`` to ``OptunaSearchCV`` and add a new sampler argument to the constructor.
> We think that ``random_state`` argument can be removed if those changes are applied.

I'll be sure to do that.

File: optuna/integration/sklearn.py [hunk 1]
```diff
@@ -0,0 +1,923 @@
...
+
+        self.refit_time_ = time() - start_time
+
+        return self
+
+    def _set_verbosity(self):
+        # type: () -> None
+
+        if self.verbose > 1:
+            logging.set_verbosity(DEBUG)
```
Comment (sile) [inline optuna/integration/sklearn.py]: I'm worried that this method call will change the global log-level of Optuna's logger.
For example, a scenario like the following can happen:
```python
import optuna

# Set loglevel to DEBUG
optuna.logging.set_verbosity(DEBUG)


# Use `OptunaSearchCV`
est = SGDClassifier(max_iter=5, tol=1e-03)
param_dist = {'alpha': distributions.LogUniformDistribution(1e-04, 1e+03)}
optuna_search = optuna.integration.OptunaSearchCV(
        est,
        param_dist,
        verbose=-1  # WARNING
)
optuna_search.fit(X, y)   # Optuna's global log-level has implicitly changed here from DEBUG to WARNING.


# Use Optuna API directly
def objective(trial):
    v = ...
    return v

study = optuna.create_study()

# No DEBUG and INFO log messages won't be outputted in the optimization.
# (This is a surprising behavior for me)
study.optimize(objective, n_trials=100)  
```

File: optuna/integration/sklearn.py [hunk 2]
```diff
@@ -0,0 +1,923 @@
...
+        return self.study_.trials_dataframe
+
+    def __init__(
+        self,
+        estimator,  # type: BaseEstimator
+        param_distributions,  # type: Mapping[str, distributions.BaseDistribution]
+        cv=5,  # type: Union[int, BaseCrossValidator, None]
+        enable_pruning=False,  # type: bool
+        error_score=np.nan,  # type: Union[str, float]
+        load_if_exists=False,  # type: bool
```
Comment (sile) [inline optuna/integration/sklearn.py]: In my feeling, this constructor takes too many arguments.
How about passing `Study` instance directly and removing related arguments from the constructor?
(i.e., removes `load_if_exists`, `pruner`, `sampler`, `storage`, `study_name`, and adds `study` with the default value `None`)

I think it will improve maintainability too.  By passing `study` instance, we will not have to modify `OptunaSearchCV` when we add more arguments (or methods) to `Study` in the future.

File: optuna/integration/sklearn.py (hunk 1 above)
Comment (Y-oHr-N) [reply optuna/integration/sklearn.py]: Yes, the global log-level will be changed certainly. I want to control log messagese with `verbose` like `GridSearchCV`, but I do not know any other way.

File: optuna/integration/sklearn.py (hunk 2 above)
Comment (Y-oHr-N) [reply optuna/integration/sklearn.py]: I removed those arguements.

Comment (sile) [reply optuna/integration/sklearn.py]: Thank you!

File: optuna/integration/sklearn.py (hunk 1 above)
Comment (sile) [reply optuna/integration/sklearn.py]: Hmm.
I also don't know the perfect answer but, IMO, the problem can be mitigated by restoring the old log level before returning from `fit` call.

Comment (Y-oHr-N) [reply optuna/integration/sklearn.py]: I implemented it as you suggested.

File: optuna/integration/sklearn.py [hunk 3]
```diff
@@ -0,0 +1,925 @@
...
+                ) for key, value in fit_params.items()
+            }
+
+        classifier = is_classifier(self.estimator)
+        cv = check_cv(self.cv, y_res, classifier)
+
+        self.n_splits_ = cv.get_n_splits(X_res, y_res, groups=groups_res)
+        self.scorer_ = check_scoring(self.estimator, scoring=self.scoring)
+
+        if self.study is None:
```
Comment (g-votte) [inline optuna/integration/sklearn.py]: How about moving study initialization to `__init__`, so that the user can call `set_user_attr` before fitting.

File: optuna/integration/sklearn.py [hunk 4]
```diff
@@ -0,0 +1,925 @@
...
+
+
+def _check_sklearn_availability():
+    # type: () -> None
+
+    if not _available:
+        raise ImportError(
+            'scikit-learn is not available. Please install scikit-learn to '
+            'use this feature. scikit-learn can be installed by executing '
+            '`$ pip install scikit-learn>=0.20.0`. For further information, '
```
Comment (g-votte) [inline optuna/integration/sklearn.py]: Would you mind checking the sklearn version in `sklearn.py`?

FYI: See also an example of version check in the visualization module.
https://github.com/pfnet/optuna/blob/master/optuna/visualization.py#L107-L112

File: optuna/integration/sklearn.py [hunk 5]
```diff
@@ -0,0 +1,925 @@
...
+
+        if self.enable_pruning and not hasattr(self.estimator, 'partial_fit'):
+            raise ValueError('estimator must support partial_fit.')
+
+        if self.max_iter <= 0:
+            raise ValueError(
+                'max_iter must be > 0, got {}.'.format(self.max_iter)
+            )
+
+        if self.study is not None and self.study.direction != 'maximize':
```
Comment (g-votte) [inline optuna/integration/sklearn.py]: Please replace `'maximize'` with `structs.StudyDirection.MAXIMIZE`. Otherwise, the following case fails.

```python
    study = optuna.create_study(direction='maximize')
    optuna_search = integration.OptunaSearchCV(
        ...
        study=study,
    )

    optuna_search.fit(X, y)
```

Comment (Y-oHr-N) [reply optuna/integration/sklearn.py]: done.

File: optuna/integration/sklearn.py [hunk 6]
```diff
@@ -0,0 +1,925 @@
...
+    indices  # type: OneDimArrayLikeType
+):
+    # type: (...) -> Union[OneDimArrayLikeType, TwoDimArrayLikeType]
+    if X is None:
+        return X
+
+    return sklearn_safe_indexing(X, indices)
+
+
+class Objective(object):
```
[the remark this answers is not in the document]
Comment (Y-oHr-N) [reply optuna/integration/sklearn.py]: Done.

File: optuna/integration/sklearn.py (hunk 3 above)
Comment (g-votte) [reply optuna/integration/sklearn.py]: Let's forget about this matter for this PR.

According to the sklearn contribution guide, any parameter validation should not be done in `__init__`.
